File: shared/src/commonTest/kotlin/io/github/yuanbaobaao/petallink/sync/PlannerTest.kt
package io.github.yuanbaobaao.petallink.sync

import io.github.yuanbaobaao.petallink.drive.DriveFile
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Planner 单测（对标 src/sync/planner.rs decide() 决策表）。
 */
class PlannerTest {
    @Test
    fun 同步状态持久化数值与原Tauri完全一致() {
        assertEquals(0, SyncStatus.SYNCED)
        assertEquals(1, SyncStatus.CLOUD_ONLY)
        assertEquals(2, SyncStatus.LOCAL_ONLY)
        assertEquals(3, SyncStatus.SYNCING)
        assertEquals(4, SyncStatus.FAILED)
        assertEquals(5, SyncStatus.CONFLICT)
        assertEquals(7, SyncStatus.DELETED)
    }

    private val path = "doc.txt"
    private fun localFile(mtime: Long = 100, size: Long = 50) =
        LocalEntry(path, mtime, size, isPlaceholder = false, isFolder = false)
    private fun cloudFile(id: String = "cid", folder: Boolean = false, editedTime: String? = null) =
        DriveFile(id = id, name = path, mimeType = if (folder) "vnd.huawei-apps.folder" else "text/plain", editedTime = editedTime)
    private fun dbEntry(
        fileId: String = "cid", localMtime: Long? = 100, localSize: Long? = 50,
        cloudTime: Long? = 1000, status: Int = SyncStatus.SYNCED, folder: Boolean = false,
    ) = DbBaselineEntry(fileId, localMtime, localSize, cloudTime, status, folder)

    @Test
    fun 全缺席返回null() {
        val result = Planner.decide(path, null, null, null, true, false)
        assertNull(result)
    }

    @Test
    fun 三方都存在_未变化返回null() {
        val result = Planner.decide(path, localFile(), cloudFile(), dbEntry(), true, false)
        assertNull(result)
    }

    @Test
    fun 三方都存在_仅本地改_上传() {
        val result = Planner.decide(path, localFile(mtime = 200), cloudFile(), dbEntry(), true, false)
        assertEquals(SyncActionType.UPLOAD, result?.type)
    }

    @Test
    fun 三方都存在_本地和云端都改_冲突() {
        val local = localFile(mtime = 200)
        val cloud = cloudFile(editedTime = "1970-01-01T00:00:02Z")
        val db = dbEntry(localMtime = 100, cloudTime = 1000)
        val result = Planner.decide(path, local, cloud, db, true, false)
        assertEquals(SyncActionType.CREATE_CONFLICT_COPY, result?.type)
    }

    @Test
    fun pending占位项收敛为Skip带cloudFile() {
        val db = dbEntry(fileId = "pending:abc")
        val result = Planner.decide(path, localFile(), cloudFile(), db, true, false)
        assertEquals(SyncActionType.SKIP, result?.type)
        assertEquals("cid", result?.cloudFile?.id)  // 携带真实 cloud_file
    }

    @Test
    fun 本地新文件_无DB_上传() {
        val result = Planner.decide(path, localFile(), null, null, true, false)
        assertEquals(SyncActionType.UPLOAD, result?.type)
        assertNull(result?.fileId)
    }

    @Test
    fun 云端新文件_无DB_创建占位() {
        val result = Planner.decide(path, null, cloudFile(), null, true, false)
        assertEquals(SyncActionType.CREATE_PLACEHOLDER, result?.type)
    }

    @Test
    fun 云端删除_本地无改_删除本地() {
        val db = dbEntry()
        val result = Planner.decide(path, localFile(), null, db, true, false)
        assertEquals(SyncActionType.DELETE_FROM_LOCAL, result?.type)
    }

    @Test
    fun 云端删除_本地有改_备份() {
        val db = dbEntry(localMtime = 100)
        val local = localFile(mtime = 200)  // 本地已改
        val result = Planner.decide(path, local, null, db, true, false)
        assertEquals(SyncActionType.BACKUP_BEFORE_CLOUD_DELETE, result?.type)
    }

    @Test
    fun 本地删除_会话内_删云端() {
        val db = dbEntry()
        val result = Planner.decide(path, null, cloudFile(), db, true, false)
        assertEquals(SyncActionType.DELETE_FROM_CLOUD, result?.type)
    }

    @Test
    fun 本地删除_启动恢复_跳过() {
        val db = dbEntry()
        val result = Planner.decide(path, null, cloudFile(), db, true, isStartupResume = true)
        // 启动恢复 + 非 DELETED → 重建占位
        assertEquals(SyncActionType.CREATE_PLACEHOLDER, result?.type)
    }

    @Test
    fun 本地删除_启动恢复_DELETED墓碑_跳过() {
        val db = dbEntry(status = SyncStatus.DELETED)
        val result = Planner.decide(path, null, cloudFile(), db, true, isStartupResume = true)
        assertEquals(SyncActionType.SKIP, result?.type)
    }

    @Test
    fun 孤儿占位符_清理() {
        val local = LocalEntry(path, 100, 0, isPlaceholder = true, isFolder = false)
        val result = Planner.decide(path, local, null, null, true, false)
        assertEquals(SyncActionType.DELETE_FROM_LOCAL, result?.type)
    }

    @Test
    fun pending加FAILED_不重试() {
        val db = dbEntry(fileId = "pending:abc", status = SyncStatus.FAILED)
        val result = Planner.decide(path, localFile(), null, db, true, false)
        assertNull(result)  // FAILED → 不自动重试
    }

    @Test
    fun pending非FAILED_重新上传() {
        val db = dbEntry(fileId = "pending:abc", status = SyncStatus.PENDING)
        val result = Planner.decide(path, localFile(), null, db, true, false)
        assertEquals(SyncActionType.UPLOAD, result?.type)
    }

    @Test
    fun 不可信云端树_抑制删除() {
        val db = dbEntry()
        val result = Planner.decide(path, localFile(), null, db, cloudTreeTrusted = false, false)
        assertEquals(SyncActionType.DELETE_FROM_LOCAL, result?.type)  // decide 仍产生动作
    }

    @Test
    fun plan不可信云端树抑制删除动作() {
        val snapshot = SyncSnapshot(
            local = mapOf(path to localFile()),
            cloud = emptyMap(),
            db = mapOf(path to dbEntry()),
            cloudTreeTrusted = false,  // 不可信
            isStartupResume = false,
        )
        val actions = Planner.plan(snapshot)
        // DELETE_FROM_LOCAL 被抑制 → 空计划
        assertTrue(actions.none { it.type == SyncActionType.DELETE_FROM_LOCAL })
    }

    @Test
    fun isLocalChanged_mtime变化为true() {
        assertTrue(Planner.isLocalChanged(localFile(mtime = 200), dbEntry(localMtime = 100)))
    }

    @Test
    fun isLocalChanged_未变为false() {
        assertEquals(false, Planner.isLocalChanged(localFile(mtime = 100), dbEntry(localMtime = 100)))
    }

    @Test
    fun isCloudChanged_cloudTime缺失为false() {
        assertEquals(false, Planner.isCloudChanged(null, dbEntry(cloudTime = 1000)))
    }

    @Test
    fun editedTime_RFC3339真实参与云端变更判定() {
        val result = Planner.decide(
            path,
            localFile(),
            cloudFile(editedTime = "1970-01-01T00:00:02Z"),
            dbEntry(cloudTime = 1000),
            true,
            false,
        )
        assertEquals(SyncActionType.DOWNLOAD, result?.type)
    }
}

File: test/service/sync/planner_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:petal_link/entity/drive_file.dart';
import 'package:petal_link/entity/sync_item.dart';
import 'package:petal_link/service/mount/manager.dart';
import 'package:petal_link/service/sync/planner.dart';
import 'package:petal_link/service/sync/sync_actions.dart';
import 'package:petal_link/types/enums.dart';

/// SyncPlanner 三方 diff 全动作矩阵测试（对齐 Rust planner.rs `_decide` 决策表）。

LocalFileEntry localEntry(
  String rel, {
  int size = 100,
  int mtime = 1700000000000,
  bool isFolder = false,
  bool isPlaceholder = false,
}) {
  return LocalFileEntry(
    absolutePath: '/mnt/$rel',
    relativePath: rel,
    size: size,
    mtime: mtime,
    isFolder: isFolder,
    isPlaceholder: isPlaceholder,
  );
}

DriveFile cloudFile(
  String id,
  String name, {
  bool isFolder = false,
  int size = 100,
  int? editedMs,
  String? parentId,
}) {
  return DriveFile(
    id: id,
    name: name,
    category: isFolder ? FileCategory.Folder : FileCategory.Document,
    size: size,
    parentFolder: parentId != null ? [parentId] : null,
    editedTime: editedMs != null
        ? DateTime.fromMillisecondsSinceEpoch(editedMs, isUtc: true)
        : null,
  );
}

DbSnapshotEntry dbEntry(
  String fileId, {
  int? localMtime = 1700000000000,
  int? localSize = 100,
  int? cloudEditedTime = 1700000000000,
  SyncItemStatus status = SyncItemStatus.Synced,
  bool isFolder = false,
}) {
  return DbSnapshotEntry(
    fileId: fileId,
    localMtime: localMtime,
    localSize: localSize,
    cloudEditedTime: cloudEditedTime,
    status: status,
    isFolder: isFolder,
  );
}

SyncSnapshot snap({
  Map<String, LocalFileEntry>? local,
  Map<String, DriveFile>? cloud,
  Map<String, DbSnapshotEntry>? db,
  bool trusted = true,
  bool startup = false,
}) {
  return SyncSnapshot(
    local: local ?? {},
    cloud: cloud ?? {},
    db: db ?? {},
    cloudTreeTrusted: trusted,
    isStartupResume: startup,
  );
}

void main() {
  final planner = SyncPlanner();

  group('三方都存在', () {
    test('双方未变化 → 无动作', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions, isEmpty);
    });

    test('本地变更 → Upload', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt', mtime: 1700000060000)},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.upload);
      expect(actions.single.fileId, 'f1');
    });

    test('本地 size 变更（mtime 相同）→ Upload（v3 精度兜底）', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt', size: 200)},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.upload);
    });

    test('云端变更 → Download', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000060000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.download);
      expect(actions.single.cloudFile?.id, 'f1');
    });

    test('双端均变更 → CreateConflictCopy', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt', mtime: 1700000060000)},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000060000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.createConflictCopy);
    });

    test('pending: 占位 + 云端已有 → Skip 携带 cloudFile（收敛为已同步）', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('$pendingFileIdPrefix${'a.txt'}')},
      ));
      expect(actions.single.actionType, SyncActionType.skip);
      expect(actions.single.cloudFile?.id, 'f1');
    });

    test('无 DB 记录 → Skip（无 cloudFile 被过滤，交给 reconcile）', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
      ));
      expect(actions, isEmpty);
    });
  });

  group('本地有 + 云端无', () {
    test('本地新文件 → Upload（无 fileId）', () {
      final actions = planner.plan(snap(
        local: {'new.txt': localEntry('new.txt')},
      ));
      expect(actions.single.actionType, SyncActionType.upload);
      expect(actions.single.fileId, isNull);
    });

    test('本地新目录 → CreateFolder', () {
      final actions = planner.plan(snap(
        local: {'dir': localEntry('dir', isFolder: true)},
      ));
      expect(actions.single.actionType, SyncActionType.createFolder);
      expect(actions.single.fileId, isNull);
    });

    test('孤儿占位符 → DeleteFromLocal（无 fileId）', () {
      final actions = planner.plan(snap(
        local: {'ghost.txt': localEntry('ghost.txt', size: 0, isPlaceholder: true)},
      ));
      expect(actions.single.actionType, SyncActionType.deleteFromLocal);
      expect(actions.single.fileId, isNull);
    });

    test('云端已删 + 本地未改 → DeleteFromLocal', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.deleteFromLocal);
      expect(actions.single.fileId, 'f1');
    });

    test('云端已删 + 本地已改 → BackupBeforeCloudDelete', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt', mtime: 1700000060000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.backupBeforeCloudDelete);
    });

    test('云端已删目录 → DeleteFromLocal（engine 决定是否保留）', () {
      final actions = planner.plan(snap(
        local: {'dir': localEntry('dir', isFolder: true)},
        db: {'dir': dbEntry('f1', isFolder: true)},
      ));
      expect(actions.single.actionType, SyncActionType.deleteFromLocal);
    });

    test('pending: 占位 + 云端无 → 重新 Upload', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        db: {'a.txt': dbEntry('$pendingFileIdPrefix${'a.txt'}')},
      ));
      expect(actions.single.actionType, SyncActionType.upload);
      expect(actions.single.fileId, isNull);
    });

    test('pending: 占位 + FAILED → 无动作（等手动重试）', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        db: {
          'a.txt': dbEntry('$pendingFileIdPrefix${'a.txt'}',
              status: SyncItemStatus.Failed),
        },
      ));
      expect(actions, isEmpty);
    });

    test('启动恢复期 + 本地未改 → Skip（删除守卫，被过滤）', () {
      final actions = planner.plan(snap(
        local: {'a.txt': localEntry('a.txt')},
        db: {'a.txt': dbEntry('f1')},
        startup: true,
      ));
      expect(actions, isEmpty);
    });
  });

  group('本地无 + 云端有', () {
    test('云端新文件 → CreatePlaceholder', () {
      final actions = planner.plan(snap(
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
      ));
      expect(actions.single.actionType, SyncActionType.createPlaceholder);
      expect(actions.single.fileId, 'f1');
    });

    test('会话内删除 → DeleteFromCloud', () {
      final actions = planner.plan(snap(
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions.single.actionType, SyncActionType.deleteFromCloud);
    });

    test('启动恢复期 + tombstone → 跳过（不重建）', () {
      final actions = planner.plan(snap(
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1', status: SyncItemStatus.Deleted)},
        startup: true,
      ));
      expect(actions, isEmpty);
    });

    test('启动恢复期 + 有 DB 非 tombstone → CreatePlaceholder（重建占位）', () {
      final actions = planner.plan(snap(
        cloud: {'a.txt': cloudFile('f1', 'a.txt', editedMs: 1700000000000)},
        db: {'a.txt': dbEntry('f1')},
        startup: true,
      ));
      expect(actions.single.actionType, SyncActionType.createPlaceholder);
    });

    test('云端文件夹 + 本地缺失 + 无 DB → CreateFolder', () {
      final actions = planner.plan(snap(
        cloud: {'dir': cloudFile('f1', 'dir', isFolder: true)},
      ));
      expect(actions.single.actionType, SyncActionType.createFolder);
      expect(actions.single.cloudFile?.id, 'f1');
    });

    test('云端文件夹 + 会话内本地删除 → DeleteFromCloud', () {
      final actions = planner.plan(snap(
        cloud: {'dir': cloudFile('f1', 'dir', isFolder: true)},
        db: {'dir': dbEntry('f1', isFolder: true)},
      ));
      expect(actions.single.actionType, SyncActionType.deleteFromCloud);
    });

    test('双方都有文件夹 → 无动作', () {
      final actions = planner.plan(snap(
        local: {'dir': localEntry('dir', isFolder: true)},
        cloud: {'dir': cloudFile('f1', 'dir', isFolder: true)},
        db: {'dir': dbEntry('f1', isFolder: true)},
      ));
      expect(actions, isEmpty);
    });
  });

  group('可信边界', () {
    test('云端不可信 → 抑制双向删除动作', () {
      final actions = planner.plan(snap(
        local: {'gone.txt': localEntry('gone.txt')},
        cloud: {'c.txt': cloudFile('f2', 'c.txt', editedMs: 1700000000000)},
        db: {
          'gone.txt': dbEntry('f1'),
          'c.txt': dbEntry('f2'),
        },
        trusted: false,
      ));
      // DeleteFromLocal（gone.txt）与 DeleteFromCloud（c.txt）都被抑制
      expect(
          actions.where((a) =>
              a.actionType == SyncActionType.deleteFromLocal ||
              a.actionType == SyncActionType.deleteFromCloud),
          isEmpty);
    });

    test('云端不可信 → 上传/占位等非删除动作保留', () {
      final actions = planner.plan(snap(
        local: {'new.txt': localEntry('new.txt')},
        cloud: {'c.txt': cloudFile('f2', 'c.txt', editedMs: 1700000000000)},
        trusted: false,
      ));
      expect(actions.map((a) => a.actionType),
          containsAll([SyncActionType.upload, SyncActionType.createPlaceholder]));
    });
  });

  group('双方都删了', () {
    test('本地无 + 云端无 + DB 有 → 无动作（engine 周期末清残余）', () {
      final actions = planner.plan(snap(
        db: {'a.txt': dbEntry('f1')},
      ));
      expect(actions, isEmpty);
    });
  });
}
